### PR TITLE
Enable Module Template

### PR DESCRIPTION
**Pull request by ElliotFriedman**
Create a new template to enable modules.

Not currently ready for review.

Should close https://github.com/ethereum-optimism/superchain-ops/issues/621

### CODE REVIEW

File: src/improvements/template/EnableDeputyPauseModuleTemplate.sol [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+        string[] memory storageWrites;
+
+        if (CHAIN_ID == MAINNET_CHAIN_ID) {
+            storageWrites = new string[](2);
+            storageWrites[1] = "LivenessGuard";
+        } else {
+            // otherwise, we only need to write to the Safe
+            storageWrites = new string[](1);
+        }
```
Comment (mds1) [inline src/improvements/template/EnableDeputyPauseModuleTemplate.sol:72]: Sepolia also has the Liveness Guard
```suggestion
        if (CHAIN_ID == getChain("mainnet").chainId || getChain("sepolia").chainId) {
            storageWrites = new string[](2);
            storageWrites[1] = "LivenessGuard";
        } else {
            // otherwise, we only need to write to the Safe
            storageWrites = new string[](1);
        }
```

File: src/improvements/template/EnableDeputyPauseModuleTemplate.sol [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+        storageWrites[0] = _SAFE_ADDRESS;
```
Comment (mds1) [inline src/improvements/template/EnableDeputyPauseModuleTemplate.sol:74]: readability nit, seems weird to write array index 0 after 1, I'd move this line to be right after we declare `string[] memory storageWrites;`

File: src/improvements/template/EnableDeputyPauseModuleTemplate.sol [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+        return storageWrites;
+    }
+
+    /// @notice Sets up the template with module configuration from a TOML file
+    /// @param taskConfigFilePath Path to the TOML configuration file
+    function _templateSetup(string memory taskConfigFilePath) internal override {
+        string memory file = vm.readFile(taskConfigFilePath);
+        newModule = vm.parseTomlAddress(file, ".newModule");
+        assertNotEq(newModule.code.length, 0, "new module must have code");
```
Comment (mds1) [inline src/improvements/template/EnableDeputyPauseModuleTemplate.sol:84]: We can strength the pre-check assertions here to make sure the deuputy pause module is configured correctly, instead of just checking code:

```solidity
require(
  IDeputyPauseModule(IDeputyPauseModule(newModule).foundationSafe() == addrRegistry.getAddress(safeAddressString),
  "foundation safe mismatch"
);
require(
  IDeputyPauseModule(IDeputyPauseModule(newModule).superchainConfig() == addrRegistry.getAddress("SuperchainConfig"),
  "SuperchainConfig mismatch"
);
```

File: src/improvements/template/EnableDeputyPauseModuleTemplate.sol [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+        // only allow one chain to be modified at a time with this template
+        AddressRegistry.ChainInfo[] memory _chains =
+            abi.decode(vm.parseToml(vm.readFile(taskConfigFilePath), ".l2chains"), (AddressRegistry.ChainInfo[]));
```
Comment (mds1) [inline src/improvements/template/EnableDeputyPauseModuleTemplate.sol:88]: This is a "superchain level" task, i.e. we are applying it directly to a safe, not to protocol contracts. So I don't think this is the right framing or config file structure (i.e. there are no L2 chains to specify here)

File: src/improvements/template/EnableDeputyPauseModuleTemplate.sol (hunk 3 above)
Comment (mds1) [reply src/improvements/template/EnableDeputyPauseModuleTemplate.sol:84]: Ok, I see you have these in the validate method. Putting them here technically makes more sense to me because we are verifying the module is configured correctly before installing it, but that doesn't matter much in this case

File: src/improvements/template/EnableDeputyPauseModuleTemplate.sol [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+                            sentinelModuleValue,
+                            bytes32(uint256(uint160(newModule))),
+                            "sentinel does not point to new module"
+                        );
+
+                        moduleWriteFound = true;
+                    }
+                }
+            }
+        }
```
Comment (mds1) [inline src/improvements/template/EnableDeputyPauseModuleTemplate.sol:176]: A lot of this can replaced with the account access parser library! Specifically it's `getStateDiffFor` method. Also, this should logic should just be in `_validate` — `checkStateDiff` is intended to have a default implementation that is never changed, but I see `_validate` does not yet take account accesses as an input

File: src/improvements/template/EnableDeputyPauseModuleTemplate.sol [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        assertTrue(moduleWriteFound, "Module write not found");
+    }
+}
+
+interface IDeputyGuardianModuleFetcher {
+    function deputyGuardian() external view returns (address);
+    function foundationSafe() external view returns (address);
+    function superchainConfig() external view returns (address);
+    function version() external view returns (string memory);
+}
```
Comment (mds1) [inline src/improvements/template/EnableDeputyPauseModuleTemplate.sol:188]: We can delete this and just import the actual interface from `lib/optimism/packages/contracts-bedrock/interfaces/safe/IDeputyPauseModule.sol`

File: test/tasks/Regression.t.sol [hunk 7]
```diff
@@ -110,6 +111,24 @@ contract RegressionTest is Test {
         assertEq(keccak256(bytes(dataToSign)), keccak256(bytes(expectedDataToSign)));
     }
 
+    function testRegressionCallDataMatches_EnableDeputyPauseModuleTemplate() public {
```
Comment (mds1) [inline test/tasks/Regression.t.sol:114]: Is the developer UX that part of template creation requires adding a corresponding regression test? That seems reasonable to me if so, but just want to confirm

File: src/improvements/template/EnableDeputyPauseModuleTemplate.sol (hunk 1 above)
Comment (mds1) [reply src/improvements/template/EnableDeputyPauseModuleTemplate.sol:72]: Ah right the DPM goes on a foundation safe, which does not have the liveness guard on both mainnet and sepolia. So I think we can delete this altogether?

File: src/improvements/template/EnableDeputyPauseModuleTemplate.sol (hunk 2 above)
Comment (mds1) [reply src/improvements/template/EnableDeputyPauseModuleTemplate.sol:74]: yea the alternative is setting the length before the if based on the conditional you already have, but doesn't matter, so resolving

File: src/improvements/template/EnableDeputyPauseModuleTemplate.sol (hunk 4 above)
Comment (mds1) [reply src/improvements/template/EnableDeputyPauseModuleTemplate.sol:88]: So the guidance for devs is: If generating a generic task that doesn't apply so a specific chain, just use the L1 chain ID? Does that have any consequences or side effects? Let's make sure to include this in the docs

File: src/improvements/template/EnableDeputyPauseModuleTemplate.sol (hunk 5 above)
Comment (mds1) [reply src/improvements/template/EnableDeputyPauseModuleTemplate.sol:176]: > should `_validate` take an account accesses parameter?

Yes! That is how you validate arbitrary aspects of the returned state diff. `validate` UX should analogous to our current `postCheck`, and `checkStateDiff` remains unchanged (i.e. devs never need to think about `checkStateDiff`, its always run, automatically)

File: test/tasks/Regression.t.sol (hunk 7 above)
Comment (mds1) [reply test/tasks/Regression.t.sol:114]: Sanity checking we are working on "template creation docs" that explain this? It also looks like these tests duplicate tons of code, we probably can make it a lot easier to implement and maintain these tests with some helper functions

File: src/improvements/tasks/MultisigTask.sol [hunk 8]
```diff
@@ -927,7 +927,7 @@ abstract contract MultisigTask is Test, Script {
     /// check the state changes applied by the task. This function can check
     /// that only the nonce changed in the parent multisig when executing a task
     /// by checking the slot and address where the slot changed.
-    function checkStateDiff(VmSafe.AccountAccess[] memory accountAccesses) internal view virtual {
+    function checkStateDiff(VmSafe.AccountAccess[] memory accountAccesses) internal virtual {
```
Comment (blmalone) [inline src/improvements/tasks/MultisigTask.sol:930]: Can this function not remain view? I've made the change locally and it seems to work.

File: src/improvements/tasks/MultisigTask.sol [hunk 9]
```diff
@@ -144,7 +144,7 @@ abstract contract MultisigTask is Test, Script {
 
     /// @notice abstract function to be implemented by the inheriting contract
     /// specifies the addresses that must have their storage written to
-    function _taskStorageWrites() internal pure virtual returns (string[] memory);
+    function _taskStorageWrites() internal virtual returns (string[] memory);
```
Comment (blmalone) [inline src/improvements/tasks/MultisigTask.sol:147]: can we keep this pure?

File: src/improvements/template/EnableDeputyPauseModuleTemplate.sol (hunk 4 above)
Comment (blmalone) [reply src/improvements/template/EnableDeputyPauseModuleTemplate.sol:88]: +1 to this, having l2chains included for this is a counter intuitive flow. We should also have the same flow for OPCM tasks as this is also a "superchain level" task.

File: src/improvements/tasks/MultisigTask.sol (hunk 8 above)
Comment (blmalone) [reply src/improvements/tasks/MultisigTask.sol:930]: I don't think we want this to be virtual either. Templates won't be overriding this functionality. It should be run all the time regardless.

File: test/tasks/mock/configs/EnableDeputyPauseModuleTemplate.toml [hunk 10]
```diff
@@ -0,0 +1,8 @@
+# this is the file used to determine the network configuration
+
+l2chains = [{name = "OP Testnet", chainId = 11155420}]
```
Comment (blmalone) [inline test/tasks/mock/configs/EnableDeputyPauseModuleTemplate.toml:3]: Can we add an example task under `src/improvements/tasks/example/sep`? My PR will pick this up later and simulate it on every CI run.

File: src/improvements/tasks/MultisigTask.sol (hunk 8 above)
Comment (mds1) [reply src/improvements/tasks/MultisigTask.sol:930]: It is currently virtual in the legacy code, but agreed it needs to be

File: src/improvements/template/EnableDeputyPauseModuleTemplate.sol (hunk 4 above)
Comment (blmalone) [reply src/improvements/template/EnableDeputyPauseModuleTemplate.sol:88]: I'm fine with it being a separate PR.

File: src/improvements/template/EnableDeputyPauseModuleTemplate.sol [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+    /// @param chainId The ID of the L2 chain
+    function _buildPerChain(uint256 chainId) internal override {}
+
+    /// @notice Builds the action for enabling the module in the Safe
+    function _buildSingle() internal override {
+        ModuleManager(parentMultisig).enableModule(newModule);
+    }
+
+    /// @notice Validates that the module was enabled correctly
+    /// @param chainId The ID of the L2 chain to validate
```
Comment (prat-gpt) [inline src/improvements/template/EnableDeputyPauseModuleTemplate.sol:77]: natspec nit

File: src/improvements/template/EnableDeputyPauseModuleTemplate.sol (hunk 4 above)
Comment (blmalone) [reply src/improvements/template/EnableDeputyPauseModuleTemplate.sol:62]: Going to resolve this issue as it is being tracked in: https://github.com/ethereum-optimism/superchain-ops/issues/396

File: src/improvements/template/EnableDeputyPauseModuleTemplate.sol (hunk 3 above)
Comment (blmalone) [reply src/improvements/template/EnableDeputyPauseModuleTemplate.sol:58]: Going to resolve this thread. I think @mds1 raises a good point, having the ability to do prevalidations should be possible. I'm opening an issue for this to look into it further.

Comment (blmalone) [reply src/improvements/template/EnableDeputyPauseModuleTemplate.sol:58]: Made this issue: https://github.com/ethereum-optimism/superchain-ops/issues/654

File: src/improvements/template/EnableDeputyPauseModuleTemplate.sol (hunk 5 above)
Comment (blmalone) [reply src/improvements/template/EnableDeputyPauseModuleTemplate.sol]: Going to resolve this thread as the commit above looks like it addresses Solomons comment.

File: test/tasks/Regression.t.sol (hunk 7 above)
Comment (blmalone) [reply test/tasks/Regression.t.sol:114]: The commit above doesn't include a refactor of `Regression.t.sol` with helper functions to avoid code duplication. I've raised an issue for this to be tackled later: https://github.com/ethereum-optimism/superchain-ops/issues/655

Comment (blmalone) [reply test/tasks/Regression.t.sol:114]: Resolving.

File: src/improvements/template/EnableDeputyPauseModuleTemplate.sol (hunk 6 above)
Comment (blmalone) [reply src/improvements/template/EnableDeputyPauseModuleTemplate.sol]: Resolving.

File: test/tasks/mock/MockMultisigTask.sol [hunk 12]
```diff
@@ -48,7 +48,7 @@ contract MockMultisigTask is MultisigTask {
         }
```
Comment (blmalone) [inline test/tasks/mock/MockMultisigTask.sol:48]: The build is failing here, we need this import: `import {VmSafe} from "forge-std/Vm.sol";`

**Review by blmalone — COMMENTED** [review]
build is failing

File: src/improvements/template/EnableDeputyPauseModuleTemplate.sol [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+        );
+
+        bytes32 moduleSlot = keccak256(abi.encode(newModule, MODULE_MAPPING_STORAGE_OFFSET));
+        bytes32 sentinelSlot = keccak256(abi.encode(SENTINEL_MODULE, MODULE_MAPPING_STORAGE_OFFSET));
+
+        bool moduleWriteFound;
+
+        address[] memory uniqueWrites = accountAccess.getUniqueWrites();
+        assertEq(uniqueWrites.length, 1, "should only write to foundation ops safe");
+        assertEq(uniqueWrites[0], parentMultisig, "should only write to foundation ops safe address");
```
[the remark this answers is not in the document]
Comment (blmalone) [reply src/improvements/template/EnableDeputyPauseModuleTemplate.sol:111]: `uniqueWrites[0]` is `0xfb1bffC9d739B8D520DaF37dF666da4C687191EA` which is the same as the GnosisSafeL2 predeploy: https://github.com/safe-global/safe-smart-account/blob/a1e7f4a763952f5b116a8ab4c7361b95de2083c3/CHANGELOG.md?plain=1#L444

[the remark this answers is not in the document]
Comment (blmalone) [reply src/improvements/template/EnableDeputyPauseModuleTemplate.sol:112]: Closing as this was fixed in: https://github.com/ethereum-optimism/superchain-ops/pull/656

**Review by blmalone — APPROVED** [review]
